src/Lecture2.hs

```diff
@@ -7,8 +7,316 @@ Stability               : Stable
 Portability             : Portable
 
 Exercises for the Lecture 2 of the Haskell Beginners course.
+
+As in the previous section, implement functions and provide type
+signatures. If the type signature is not already written, write the
+most polymorphic type signature you can.
+
+Unlike exercises to Lecture 1, this module also contains more
+challenging exercises. You don't need to solve them to finish the
+course but you can if you like challenges :)
 -}
 
 module Lecture2
-    (
+    ( -- * Normal
+      lazyProduct
+    , duplicate
+    , removeAt
+    , evenLists
+    , dropSpaces
+
+    , Knight (..)
+    , dragonFight
+
+      -- * Hard
+    , isIncreasing
+    , merge
+    , mergeSort
+
+    , Expr (..)
+    , Variables
+    , EvalError (..)
+    , eval
+    , constantFolding
     ) where
+
+{- | Implement a function that finds a product of all the numbers in
+the list. But implement a lazier version of this function: if you see
+zero, you can stop calculating product and return 0 immediately.
+
+>>> lazyProduct [4, 3, 7]
+84
+-}
+
+-- Hi, Dmitrii. Sorry, I don't solve all exercises. I'll try to solve them later and send you PR. Thank you.
+
+lazyProduct :: [Int] -> Int
+lazyProduct = go 1
+    where
+        go :: Int -> [Int] -> Int
+        go acc [] = acc
+        go acc (x : xs)
+            | x == 0 = 0
+            | otherwise = go (acc * x)  xs
+
+{- | Implement a function that duplicates every element in the list.
+
+>>> duplicate [3, 1, 2]
+[3,3,1,1,2,2]
+>>> duplicate "cab"
+"ccaabb"
+-}
+duplicate :: [a] -> [a]
+duplicate [] = []
+duplicate (x : xs) = x : x : duplicate xs
+
+{- | Implement function that takes index and a list and removes the
+element at the given position. Additionally, this function should also
+return the removed element.
+
+>>> removeAt 0 [1 .. 5]
+(Just 1,[2,3,4,5])
+
+>>> removeAt 10 [1 .. 5]
+(Nothing,[1,2,3,4,5])
+-}
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt _ [] = (Nothing, [])
+removeAt 0 (x : xs) = (Just x, xs)
+removeAt n list = case splitAt n list of
+    ([], after) -> (Nothing, after)
```

Comment:
That's a neat trick 👍🏻 
I guess you're pattern-matching here to handle cases with negative numbers 🙂

Reply:
apparently not )